Specify end of stream behaviour (#24)
cabal-club/cable: handshake.md

Incoming Cable Wire Protocol messages will also be length-prefixed, and the
message bodies will be encrypted as *ciphertext*s, and must be run through
Noise for decryption. There are additional steps to handle message

> fragmentation and end of stream markers, described in the next subsection.

Reviewer: ```suggestion
fragmentation and the end of the stream, described in the next subsection.
```

3. `WriteMsg(plaintext)`

> 4. `WriteBytes(0)`

Reviewer: ```suggestion
```

The stream isn't terminated at the end of each message, so I don't believe this belongs here.

If, for example, a `plaintext` of length 90200 were to be encoded & written,
the first 65519 bytes would first be encrypted and written, followed by the
remaining 24681 bytes being encrypted and written. Since each ciphertext has an
extra 16 bytes added of authentication data, and there are two segments
written, the total written length would be `65519 + 24681 + 16 * 2 = 90232`
bytes.

> The fourth step is described in the [End of stream](#55-end-of-stream)
> subsection.
> 

Reviewer: ```suggestion
```

3. The resulting bytes `plaintext` may then be parsed as a Cable Wire Protocol
   message.

### 5.5 End of stream
After a complete Cable Wire Protocol message is sent, regardless of whether
it consists of one or several fragments, the sender MUST write a message of

> `length=0`. This final message serves as an end of stream marker for the

Reviewer: I don't believe this is true. A host may send *many* Cable Wire Protocol messages before choosing to terminate the stream.

> ### 5.5 End of stream
> After a complete Cable Wire Protocol message is sent, regardless of whether
> it consists of one or several fragments, the sender MUST write a message of
> `length=0`. This final message serves as an end of stream marker for the
> receiving peer and enables the peer to differentiate between an intentionally
> terminated stream and a dropped connection.

Reviewer: ```suggestion
### 5.5 End of stream
When a host has decided to terminate the exchange of messages, they MUST send a message of length zero to indicate this intention, and MUST NOT send any further messages.

The host receiving a message of length zero SHOULD respond by writing back a message of length zero, to indicate it has also finished writing. An implementation SHOULD have a time-out of some kind in case the other side does not transmit an end-of-stream marker or it is truncated by an attacker.

When a host has both sent and received a length-zero message, it is then safe for the underlying transport to do its own disconnect logic, if any.
```
